mart/attack/adversary.py

```diff
@@ -131,57 +150,43 @@ def training_step(self, batch, batch_idx):
             if len(gain.shape) > 0:
                 gain = gain[~found]
 
-        if len(gain.shape) > 0:
-            gain = gain.sum()
-
-        # Log gain as a metric for LR scheduler to monitor, and show gain on progress bar.
-        self.log("gain", gain, prog_bar=True)
-
-        return gain
+        return gain.sum()
 
     def configure_gradient_clipping(
-        self, optimizer, optimizer_idx, gradient_clip_val=None, gradient_clip_algorithm=None
+        self, optimizer, gradient_clip_val=None, gradient_clip_algorithm=None
     ):
         # Configuring gradient clipping in pl.Trainer is still useful, so use it.
-        super().configure_gradient_clipping(
-            optimizer, optimizer_idx, gradient_clip_val, gradient_clip_algorithm
-        )
+        super().configure_gradient_clipping(optimizer, gradient_clip_val, gradient_clip_algorithm)
 
         if self.gradient_modifier:
             for group in optimizer.param_groups:
                 self.gradient_modifier(group["params"])
 
     @silent()
-    def forward(self, *, model=None, **batch):
-        batch["model"] = model
-
-        # Adversary can live within a sequence of model. To signal the adversary should
-        # attack, one must pass a model to attack when calling the adversary. Since we
-        # do not know where the Adversary lives inside the model, we also need the
-        # remaining sequence to be able to get a loss.
-        if model:
-            self._attack(**batch)
-
-        perturbation = self.perturber(**batch)
-        input_adv = self.composer(perturbation, **batch)
+    def forward(self, *, input, target, model):
+        # What we need is a frozen model that returns (a dictionary of) logits, or losses.
+        self.model = model
```

Comment:
This is a bad idea...

Reply:
Could you elaborate? I made an commit to hide the model in [2cd7ca8](https://github.com/IntelLabs/MART/pull/200/commits/2cd7ca86bf1f8886b6f0e110347696d5fd908736)

Reply:
Now I hide `model` in adversary's data batch.

Reply:
It's easier to reason about modules when they're purely functional. Modifying self means one needs to mentally keep track of state of the module now.